File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.cc
#include <iostream>
#include <string>
#include <stack>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    while(true) {
        string s;
        getline(cin, s);
        stack<char> t;
        bool is_balanced = true;

        if(s == ".")    break;

        for(auto temp : s){
            if(temp == '(' || temp == '['){
                t.push(temp);
            }
            else if(temp == ')'){
                if(!t.empty() && t.top() == '(')
                    t.pop();
                else{
                    is_balanced = false;
                    break;
                }
            }
            else if(temp == ']'){
                if(!t.empty() && t.top() == '[')
                    t.pop();
                else{
                    is_balanced = false;
                    break;
                }
            }
        }

        if(!t.empty())
            is_balanced = false;

        cout << (is_balanced ? "yes" : "no") << '\n';
    }

    return 0;
}

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

class Main {
    public static void main(String[] args) throws IOException {
        
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        while (true) 
        { 
            char chars[] = br.readLine().toCharArray();
            if(chars[0] == '.' & chars.length == 1)   break;
            
            Stack<Character> stackChar = new Stack<>();

            boolean res = true;

            for(char c : chars) {
                if (c == '(')   stackChar.push(c);
                else if (c == '[')  stackChar.push(c);
                else if (c == ')')  {
                    if(stackChar.empty() || stackChar.pop() != '(')  {
                        res = false;
                        break;
                    }
                }
                else if (c == ']')  {
                    if  (stackChar.empty() || stackChar.pop() != '[')    {
                        res = false;
                        break;
                    }
                }
            }
            if(!stackChar.empty())  res = false;

            if(res) System.out.println("yes");
            else    System.out.println("no");

        }
    }
}